src/ProjectTemplates/Web.ProjectTemplates/content/BlazorWeb-CSharp/BlazorWeb-CSharp/Components/Account/Shared/PasskeySubmit.razor.js

```diff
@@ -0,0 +1,80 @@
+async function fetchWithErrorHandling(url, options = {}) {
+    const response = await fetch(url, {
+        credentials: 'include',
+        ...options
+    });
+    if (!response.ok) {
+        const text = await response.text();
+        console.error(text);
+        throw new Error(`The server responded with status ${response.status}.`);
+    }
+    return response;
+}
+
+async function createCredential() {
+    const optionsResponse = await fetchWithErrorHandling('/Account/PasskeyCreationOptions', {
+        method: 'POST',
+    });
+    const optionsJson = await optionsResponse.json();
+    const options = PublicKeyCredential.parseCreationOptionsFromJSON(optionsJson);
+    return await navigator.credentials.create({ publicKey: options });
+}
+
+async function requestCredential(email) {
+    const optionsResponse = await fetchWithErrorHandling(`/Account/PasskeyRequestOptions?username=${email}`, {
+        method: 'POST',
+    });
+    const optionsJson = await optionsResponse.json();
+    const options = PublicKeyCredential.parseRequestOptionsFromJSON(optionsJson);
+    return await navigator.credentials.get({ publicKey: options });
+}
+
+customElements.define('passkey-submit', class extends HTMLElement {
+    static formAssociated = true;
+
+    connectedCallback() {
+        this.internals = this.attachInternals();
+        this.isObtainingCredentials = false;
+        this.attrs = {
+            operation: this.getAttribute('operation'),
+            name: this.getAttribute('name'),
+            emailName: this.getAttribute('email-name'),
+        };
+
+        this.internals.form.addEventListener('submit', (event) => {
+            if (event.submitter?.name === '__passkeySubmit') {
+                event.preventDefault();
+                this.obtainCredentialAndReSubmit();
+            }
+        });
```

Comment:
One thought that I had is we could maybe leverage the constraint validation API. Setup a custom validity to "Retrieving credential" or something similar, and then upon the credential promise resolving. Setting the value in the form data and removing the custom validity state.

That would save us this bit and the bit below.

Reply:
Interesting. We'd still need to know if the form is being submitted via the "passkey submit" button vs. another submit button, particularly in the case of login, in which case we'd still need a condition to decide whether to set the custom validity. And after clearing the custom validity, we'd still need to programmatically resubmit the form, right? It's possible I just need to look more into the details of the constraint validation API. I'll plan to investigate this in a follow-up, if that's fine with you.

Reply:
Don't take me too seriously here. I'm not 100% aware of the casuistic so I might have over simplified things. If it needs to be triggered by the login with a pass-key bit, then it might not make sense.